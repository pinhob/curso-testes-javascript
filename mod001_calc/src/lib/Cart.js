import find from "lodash/find";
import remove from "lodash/remove";

export default class Cart {
  items = [];

  add(item) {
    let itemToFind = { product: item.product };
    const itemAlreadyInCart = find(this.items, itemToFind);

    if (itemAlreadyInCart) {
      remove(this.items, itemToFind);
    }

    this.items.push(item);
  }

  remove(product) {
    remove(this.items, { product });
  }

  getTotal() {
    return this.items.reduce((acc, item) => acc + (item.product.price * item.quantity), 0)
  };

  summary() {
    const total = this.getTotal();
    const items = this.items;

    return {
      total,
      items,
    };
  }

  checkout() {
    const { total, items } = this.summary();

    this.items = [];

    return {
      total,
      items,
    };
  };
};
