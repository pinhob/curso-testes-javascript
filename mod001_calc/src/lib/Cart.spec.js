import Cart from "./Cart";

describe('Cart', () => {
  let cart;
  const product = {
    title: "Adidas Shoes - Men",
    price: 12599
  };

  const product2 = {
    title: "Adidas Shoes - Women",
    price: 12099
  };

  beforeEach(() => {
    cart = new Cart();
  });

  describe('getTotal()', () => {
    it('should return 0 when getTotal() is executed in a newly created instance', () => {
      expect(cart.getTotal().getAmount()).toBe(0);
    });

    it('should multiply quantity and price and receive the total amount', () => {
      const item = {
        product,
        quantity: 2,
      };

      cart.add(item);

      expect(cart.getTotal().getAmount()).toBe(25198);
    });

    it('should ensure no more than on product exists at a time', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product,
        quantity: 1,
      });

      expect(cart.getTotal().getAmount()).toBe(12599);
    });

    it('should update total when a product gets included and then removed', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      cart.remove(product);

      expect(cart.getTotal().getAmount()).toBe(12099);
    });
  });

  describe('checkout()', () => {
    it('should return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 2,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it('should return an object with the total and the list of items when summary() is called', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 2,
      });

      expect(cart.summary()).toMatchSnapshot();
      expect(cart.getTotal().getAmount()).toBeGreaterThan(0);
    });

    it('should include formatted amount in the summary', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 2,
      });

      console.log(cart.summary());
      expect(cart.summary().formatted).toEqual('R$493.96');
    });

    it('should reset the cart when checkout() is called', () => {
      cart.add({
        product: product2,
        quantity: 2,
      });

      cart.checkout();

      expect(cart.getTotal().getAmount()).toEqual(0);
    });
  });

  describe('Special conditions', () => {
    it('should apply percentage discount quantity above minimum is passed', () => {
      const condition = {
        percentage: 30,
        minimum: 2,
      };

      cart.add({
        product,
        condition,
        quantity: 3,
      });

      expect(cart.getTotal().getAmount()).toEqual(26458);
    });

    it('should apply quantity discount for even quantities', () => {
      const condition = {
        quantity: 2,
      }

      cart.add({
        product,
        condition,
        quantity: 4,
      });

      expect(cart.getTotal().getAmount()).toEqual(25198);
    });

    it('should apply quantity discount for odd quantities', () => {
      const condition = {
        quantity: 2,
      }

      cart.add({
        product,
        condition,
        quantity: 3,
      });

      expect(cart.getTotal().getAmount()).toEqual(22678);
    });

    it('should NOT apply percentage discount quantity if quantity is equal or below minimun', () => {
      const condition = {
        percentage: 30,
        minimum: 2,
      };

      cart.add({
        product,
        condition,
        quantity: 2,
      });

      expect(cart.getTotal().getAmount()).toEqual(25198);
    });

    it('should apply quantity discount when quantity is equal or below condition', () => {
      const condition = {
        quantity: 2,
      }

      cart.add({
        product,
        condition,
        quantity: 1,
      });

      expect(cart.getTotal().getAmount()).toEqual(12599);
    });

    it('should receive two or more conditions and determine/apply the best discount. First case.', () => {
      const condition = {
        percentage: 30,
        minimum: 2,
      };

      const condition2 = {
        quantity: 2,
      }

      cart.add({
        product,
        condition: [condition, condition2],
        quantity: 5,
      });

      expect(cart.getTotal().getAmount()).toEqual(37797);
    });

    it('should receive two or more conditions and determine/apply the best discount. Second case.', () => {
      const condition = {
        percentage: 80,
        minimum: 2,
      };

      const condition2 = {
        quantity: 2,
      }

      cart.add({
        product,
        condition: [condition, condition2],
        quantity: 5,
      });

      expect(cart.getTotal().getAmount()).toEqual(12599);
    });
  });
});