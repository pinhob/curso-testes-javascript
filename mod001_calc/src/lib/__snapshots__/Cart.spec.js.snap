// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 12599,
        "title": "Adidas Shoes - Men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 12099,
        "title": "Adidas Shoes - Women",
      },
      "quantity": 2,
    },
  ],
  "total": Object {
    "amount": 49396,
    "currency": "BRL",
    "precision": 2,
  },
}
`;

exports[`Cart checkout() should return an object with the total and the list of items when summary() is called 1`] = `
Object {
  "formatted": "R$493.96",
  "items": Array [
    Object {
      "product": Object {
        "price": 12599,
        "title": "Adidas Shoes - Men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 12099,
        "title": "Adidas Shoes - Women",
      },
      "quantity": 2,
    },
  ],
  "total": Object {
    "amount": 49396,
    "currency": "BRL",
    "precision": 2,
  },
}
`;
